File: frontend/src/lib/twitch-badges.ts
import { PUBLIC_API_URL } from '$env/static/public';

export type TwitchBadgeVersion = {
	image_url_1x: string;
	title: string;
};

export type TwitchBadges = Record<string, TwitchBadgeVersion>;

const badgeRequests = new Map<string, Promise<TwitchBadges>>();

export function getTwitchBadges(roomId: string): Promise<TwitchBadges> {
	const existingRequest = badgeRequests.get(roomId);
	if (existingRequest) return existingRequest;

	const url = new URL('/twitch/badges', PUBLIC_API_URL);
	url.searchParams.set('broadcaster_id', roomId);

	const request = fetch(url)
		.then((response) => {
			if (!response.ok) throw new Error('Unable to load Twitch badges');
			return response.json() as Promise<TwitchBadges>;
		})
		.catch(() => ({}));

	badgeRequests.set(roomId, request);
	return request;
}

File: backend/internal/routes/twitch_badges.go
package routes

import (
	"net/http"
	"regexp"

	"github.com/pocketbase/pocketbase/core"
	"github.com/seriousm4x/wubbl0rz-archiv/external"
)

var twitchBroadcasterIDPattern = regexp.MustCompile(`^[0-9]+$`)

// TwitchBadges proxies Twitch chat badge metadata without exposing app credentials.
func TwitchBadges(app core.App, e *core.RequestEvent) error {
	broadcasterID := e.Request.URL.Query().Get("broadcaster_id")
	if !twitchBroadcasterIDPattern.MatchString(broadcasterID) {
		return e.JSON(http.StatusBadRequest, map[string]string{
			"message": "broadcaster_id must be a numeric Twitch user ID",
		})
	}

	badges, err := external.TwitchGetChatBadges(app, broadcasterID)
	if err != nil {
		return e.JSON(http.StatusBadGateway, map[string]string{
			"message": "unable to load Twitch badges",
		})
	}

	return e.JSON(http.StatusOK, badges)
}
